src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,224 @@
+        status = StatusCodes.InternalServerError,
+        entity = HttpEntity(ContentTypes.`application/json`, errorReport.toJson.compactPrint))))
+  }
+
+
+  /*
+   * Helper Methods
+   */
+
+
+  /**
+    * General Approach
+    * 1. Define a `Source` of entity queries
+    * 2. Run the source events through a `Flow`.
+    * 3. Flow sends events (batch of entities) to a streaming output actor
+    * 4. Return a Done to the calling route when complete.
+    */
+  private def streamSingularType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, headers: IndexedSeq[String], attributeNames: Option[IndexedSeq[String]]): Future[Done] = {
+    // Akka Streams Support
+    implicit val system: ActorSystem = ActorSystem("Streaming-Entity-Exporter")
+    implicit val materializer: ActorMaterializer = ActorMaterializer()
+
+    // The output to the user
+    val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`text/plain`, entityType + ".txt")))
+
+    // The Source
+    val entityQuerySource = Source(entityQueries.toStream)
+
+    // MapAsync should preserve order.
+    val flow = Flow[EntityQuery].mapAsync(1) { query =>
+      getEntities(workspaceNamespace, workspaceName, entityType, query) map { entities =>
+        sendRowsAsChunks(streamingActorRef, query, entityQueries.size, entityType, headers, entities)
+      }
+    }
+
+    // Ignore the result - we don't need to remember anything about this operation.
+    val sink = Sink.ignore
+
+    // finally, run it:
+    entityQuerySource.via(flow).runWith(sink)
+
+  }
+
+  private def streamCollectionType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, attributeNames: Option[IndexedSeq[String]]): Future[_root_.akka.Done.type] = {
+    // The output file
+    lazy val zipFile = writeCollectionTypeZipFile(workspaceNamespace, workspaceName, entityType, entityQueries, metadata, attributeNames)
+
+    // The output to the user
+    lazy val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`application/octet-stream`, entityType + ".zip")))
+    zipFile map { f =>
+      streamingActorRef ! FirstChunk(HttpData.apply(f.byteArray), 0)
+    }
+
+    Future(Done)
+  }
+
+  private def sendRowsAsChunks(actorRef: ActorRef, query: EntityQuery, querySize: Int, entityType: String, headers: IndexedSeq[String], entities: Seq[Entity]): Unit = {
+    val rows = TSVFormatter.makeEntityRows(entityType, entities, headers)
+    val remaining = querySize - query.page + 1
+    // Send headers as the first chunk of data
+    if (query.page == 1) { actorRef ! FirstChunk(HttpData(headers.mkString("\t") + "\n"), remaining)}
+    // Send entities
+    actorRef ! NextChunk(HttpData(rows.map { _.mkString("\t") }.mkString("\n") + "\n"), remaining - 1)
```

Comment:
Would it be possible to use [TSVParser](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala) which uses [Univocity](https://github.com/uniVocity/univocity-parsers) under the hood? This might fail for certain exotic TSVs which contain tabs inside quotes, for example. Also it handles things like auto-detecting line endings (`\n` vs \r\n`).